certificate-authority/service/grpc/signCertificate.go

```diff
@@ -111,14 +106,28 @@ func (s *CertificateAuthorityServer) SignCertificate(ctx context.Context, req *p
 	if err != nil {
 		return nil, logger.LogAndReturnError(status.Errorf(codes.InvalidArgument, fmtError, err))
 	}
-	if signingRecord.GetCredential() == nil {
-		return nil, logger.LogAndReturnError(status.Errorf(codes.InvalidArgument, "cannot sign certificate: cannot create signing record"))
+	credential := signingRecord.GetCredential()
+	if credential == nil {
+		return nil, logger.LogAndReturnError(status.Errorf(codes.InvalidArgument, fmtError, errors.New("cannot create signing record")))
 	}
-	signingRecord.Credential.CertificatePem = string(cert)
-	if err := s.updateSigningRecord(ctx, signingRecord); err != nil {
+	credential.CertificatePem = string(cert)
+	replacedRecord, err := s.updateSigningRecord(ctx, signingRecord)
+	if err != nil {
 		return nil, logger.LogAndReturnError(status.Errorf(codes.InvalidArgument, fmtError, err))
 	}
 	logger.With("crt", string(cert)).Debugf("CertificateAuthorityServer.SignCertificate")
+	replacedCredential := replacedRecord.GetCredential()
+	if replacedRecord != nil {
+		err = s.store.RevokeCertificates(ctx, replacedCredential.GetIssuerId(), &store.RevocationListCertificate{
+			Serial:     replacedCredential.GetSerial(),
+			Expiration: replacedCredential.GetValidUntilDate(),
+			Revocation: time.Now().UnixNano(),
+		})
+		if err != nil {
+			// TODO: what to do here? remove the new signing record? restore the original?
```

Comment:
At first you need to add cerficate to CRL list. And when it success then you try set new certificate . It can fails when someone make parallel update. In this case you will return error. And when all steps are succeed than you return ok

Reply:
I need an additional request to mongo that way though, which I wanted to avoid, but I guess there is no way around it.